### ISSUE TITLE
Add option to avoid environment variables during initialization

### ISSUE THREAD

**Issue by minusnine** [c-ares/c-ares#810]
Hello!

In my organization's applications, Ares initialization generally happens lazily and in a multi-threaded context. Accessing environment variables is not thread safe.

 I'd like to propose adding a means to bypass initialization via environment. Specifically, I would like to add an [flag value](https://github.com/c-ares/c-ares/blob/67db56c3419f41c6aba4c2826c041a76714ce551/include/ares.h#L230) (say, `ARES_FLAG_NOINITENV`) to be able to skip the [`ares__init_by_environment`](https://github.com/c-ares/c-ares/blob/67db56c3419f41c6aba4c2826c041a76714ce551/src/lib/ares_sysconfig_files.c#L473) call in [`ares__init_by_sysconfig`](https://github.com/c-ares/c-ares/blob/67db56c3419f41c6aba4c2826c041a76714ce551/src/lib/ares_sysconfig.c#L521C12-L521C37), which is in turn called by [`ares_init_options`](https://github.com/c-ares/c-ares/blob/67db56c3419f41c6aba4c2826c041a76714ce551/src/lib/ares_init.c#L267).

Would this be acceptable or is there a different, preferred approach? 

Another option would be to add a flag to bypass each of the `LOCALDOMAIN` and `RES_OPTIONS` environment variables in `ares__init_by_environment`. This somewhat follows the [pattern set by `ARES_FLAG_NOALIASES`](https://github.com/c-ares/c-ares/blob/67db56c3419f41c6aba4c2826c041a76714ce551/docs/ares_init_options.3#L116). The downside is that if a new environment variables were added, a new flag would also be needed to maintain the ability to avoid environment variable access.

**Comment by bradh352** [comment]
Sure, make sense to have a flag to skip environment loading.  Could be considered a "risk" anyhow.  We have configurable options for anything that might have been read from the environment already, so no need to make it specify anything else.

**Comment by dimbleby** [comment]
> Accessing environment variables is not thread safe.

Not sure what you mean by this?  The usual interpretation is that _setting_ your own environment variables is not thread safe.  If you are seeing thread-safety problems then a simpler solution is likely "don't do that".

(There might be other reasons to introduce such an option anyway)

**Comment by dimbleby** [comment]
Right, but your problem is not that c-ares reads the environment, it is that something else is writing it. 

If maintainer is happy to accept a merge request to make this change then I have no quarrel with that.  I'm just pointing out that the case you are making for it is not convincing!

**Comment by bradh352** [comment]
no problems on this end, I'm happy to accept a PR to add an option to disable reading from the environment as long as it covers all cases.  I've never been a fan of a library using environment variables, the host application may never realize its occurring and creates hard to debug situations.  So while your interest is in thread safety, it definitely would extend beyond that to other use-cases.
